compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs

```diff
@@ -87,51 +98,44 @@ createProjectPackageDb projectRoot (disableScenarioService -> opts) modulePrefix
-              <*> (fst <$> useE GeneratePackageMap projectRoot)
-      (stablePkgs, PackageMap dependenciesInPkgDb) <- maybe (fail "Failed to generate package info") pure mbRes
       let stablePkgIds :: Set LF.PackageId
           stablePkgIds = Set.fromList $ map LF.dalfPackageId $ MS.elems stablePkgs
-      let dependenciesInPkgDbIds =
-              Set.fromList $ map LF.dalfPackageId $ MS.elems dependenciesInPkgDb
+      let baseDependenciesIds =
+              Set.fromList $ map LF.dalfPackageId $ MS.elems baseDependencies
+
+      let decodeDalf_ dalf = do
+            bs <- BS.readFile dalf
+            (dalf,) <$> either fail pure (decodeDalf baseDependenciesIds dalf bs)
 
       -- This is only used for unit-id collision checks and dependencies on newer LF versions.
       dalfsFromDependencyFps <- queryDalfs (Just [depMarker]) depsDir
-      dalfsFromDependencies <- forM dalfsFromDependencyFps $ decodeDalf_ dependenciesInPkgDbIds
       dalfsFromDataDependencyFps <- queryDalfs (Just [dataDepMarker]) depsDir
-      dalfsFromDataDependencies <- forM dalfsFromDataDependencyFps $ decodeDalf_ dependenciesInPkgDbIds
       mainDalfFps <- queryDalfs (Just [mainMarker]) depsDir
-      mainDalfs <- forM mainDalfFps $ decodeDalf_ dependenciesInPkgDbIds
 
-      let dependencyInfo = DependencyInfo
-              { dependenciesInPkgDb
-              , dalfsFromDependencies
-              , dalfsFromDataDependencies
-              , mainUnitIds = map decodedUnitId mainDalfs
-              }
+      dalfsFromDependenciesWithFps <- mapM decodeDalf_ (dalfsFromDependencyFps \\ dalfsFromDataDependencyFps)
+      dalfsFromDataDependenciesWithFps <- mapM decodeDalf_ dalfsFromDataDependencyFps
+      mainDalfsWithFps <- mapM decodeDalf_ mainDalfFps
 
-      -- We perform these check before checking for unit id collisions
-      -- since it provides a more useful error message.
-      whenLeft
-          (checkForInconsistentLfVersions (optDamlLfVersion opts) dependencyInfo)
-          exitWithError
-      whenLeft
-          (checkForIncompatibleLfVersions (optDamlLfVersion opts) dependencyInfo)
-          exitWithError
-      whenLeft
-          (checkForUnitIdConflicts dependencyInfo)
-          exitWithError
+      let
+        dalfsFromDependencies = fmap snd dalfsFromDependenciesWithFps
+        dalfsFromDataDependencies = fmap snd dalfsFromDataDependenciesWithFps
+        mainDalfs = fmap snd mainDalfsWithFps
+
+        depPkgIds = Set.fromList $ fmap (LF.dalfPackageId . decodedDalfPkg) dalfsFromDependencies
+        dataDepPkgIds = Set.fromList $ fmap (LF.dalfPackageId . decodedDalfPkg) dalfsFromDataDependencies
+
+        mainUnitIds = map decodedUnitId mainDalfs
 
       -- We run the checks for duplicate unit ids before
       -- to avoid blowing up GHC when setting up the GHC session.
-      exposedModules <- getExposedModules opts projectRoot
+      -- exposedModules <- getExposedModules opts projectRoot
```

Comment:
since this depends on the set of registered packages, I moved it down into the loop. I'm not sure about the performance implications, though